File: api-smartcard/ma3api-smartcard/csharp/src/util/SmartCardSeed.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Runtime.CompilerServices;
using System.Text;
using iaik.pkcs.pkcs11.wrapper;
using log4net;
using tr.gov.tubitak.uekae.esya.api.common;
using tr.gov.tubitak.uekae.esya.api.common.crypto;
using tr.gov.tubitak.uekae.esya.api.smartcard.pkcs11;

namespace tr.gov.tubitak.uekae.esya.api.smartcard.util
{
    public class SmartCardSeed : ISeed
    {
        private static readonly ILog logger = LogManager.GetLogger(MethodBase.GetCurrentMethod().DeclaringType);

        protected SmartCard mSC;
        private readonly long mSlotNo;
        protected long mSessionID = -1;


        public SmartCardSeed(CardType aCardType, long aSlotNo)
        {
            mSC = new SmartCard(aCardType);
            mSlotNo = aSlotNo;
            openSession();
        }


        public byte[] getSeed(int aLength)
        {
            try
            {
                return mSC.getRandomData(mSessionID, aLength);
            }
            catch (Exception e)
            {
                logger.Warn("Cannot get seed from SmartCard:" + e.Message);
                validateSession();
                try
                {
                    return mSC.getRandomData(mSessionID, aLength);
                }
                catch (Exception newEx)
                {
                    throw new ESYARuntimeException("Cannot get seed from SmartCard:" + newEx.Message, newEx);
                }
            }
        }

        [MethodImpl(MethodImplOptions.Synchronized)]
        private void validateSession() // we dont want many thread come and validate simultaneously
        { 
            CK_SESSION_INFO sessionInfo = null;
            try
            {
                sessionInfo = mSC.getSessionInfo(mSessionID);
            }
            catch (PKCS11Exception e)
            {
                logger.Info("SmartCard seed Session is Invalid:" + e.Message + " Reopening Session");
                openSession();
            }
            if (sessionInfo == null)
            {
                logger.Info("SmartCard seed Session is Null, Opening Session");
                openSession();
            }
        }

        private void openSession()
        {
            try
            {
                mSessionID = mSC.openSession(mSlotNo);
                logger.Info("SmartCard Session is opened");
            }
            catch (PKCS11Exception e)
            {
                throw new ESYARuntimeException("Error While Getting SmartCard Seed:" + e.Message, e);
            }
        }

        public SmartCard getSmartcard()
        {
            return mSC;
        }

        public long getSlotNo()
        {
            return mSlotNo;
        }

        public override bool Equals(object obj)
        {
            if (obj is SmartCardSeed)
            {
                SmartCardSeed scObj = (SmartCardSeed) obj;
                if (this.getSmartcard().getCardType().Equals(scObj.getSmartcard().getCardType()) &&
                   this.getSlotNo() == scObj.getSlotNo())
                    return true;
            }
            return false;


            
        }
    }
}

File: api-smartcard/ma3api-smartcard/src/tr/gov/tubitak/uekae/esya/api/smartcard/util/SmartCardSeed.java
package tr.gov.tubitak.uekae.esya.api.smartcard.util;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import sun.security.pkcs11.wrapper.CK_SESSION_INFO;
import sun.security.pkcs11.wrapper.PKCS11Exception;
import tr.gov.tubitak.uekae.esya.api.common.ESYARuntimeException;
import tr.gov.tubitak.uekae.esya.api.common.crypto.ISeed;
import tr.gov.tubitak.uekae.esya.api.smartcard.pkcs11.CardType;
import tr.gov.tubitak.uekae.esya.api.smartcard.pkcs11.SmartCard;
import tr.gov.tubitak.uekae.esya.api.smartcard.pkcs11.SmartCardException;

import java.io.IOException;

public class SmartCardSeed implements ISeed {

    private static final Logger logger = LoggerFactory.getLogger(SmartCardSeed.class);

    private SmartCard mSC = null;
    private long mSlotNo;
    private long mSessionID = -1;

    //static long count = 0;

    public SmartCardSeed(CardType aCardType, long aSlotNo) throws PKCS11Exception, IOException, SmartCardException {
        mSC = new SmartCard(aCardType);
        mSlotNo = aSlotNo;
        openSession();
    }

    public byte[] getSeed(int aLength) {
        try {
            return mSC.getRandomData(mSessionID, aLength);
        } catch (Exception e) {
            logger.warn("Cannot get seed from SmartCard:"+e.getMessage(), e);
            // in case we fail validate session
            validateSession();
            try {
                return mSC.getRandomData(mSessionID, aLength);
            } catch (Exception newEx) {
                throw new ESYARuntimeException("Cannot get seed from SmartCard:"+newEx.getMessage(),newEx);
            }
        }

    }

    private synchronized void validateSession() { // we dont want many thread come and validate simultaneously
        CK_SESSION_INFO sessionInfo = null;
        try {
            sessionInfo = mSC.getSessionInfo(mSessionID);
        } catch (PKCS11Exception e) {
            logger.info("SmartCard seed Session is Invalid:"+e.getMessage()+" Reopening Session", e);
            openSession();
        }
        if(sessionInfo == null){
            logger.info("SmartCard seed Session is Null, Opening Session");
            openSession();
        }
    }

    private void openSession() {
        try {
            mSessionID = mSC.openSession(mSlotNo);
            logger.info("SmartCard Session is opened");
        } catch (PKCS11Exception e) {
            logger.warn("Error While Getting SmartCard Seed. It will try again", e);
            try {
                Thread.sleep(500);
                mSessionID = mSC.openSession(mSlotNo);
            } catch (Exception ex) {
                throw new ESYARuntimeException("Error While Getting SmartCard Seed. It will not tried again:" + e.getMessage(), e);
            }
        }
    }

    public SmartCard getSmartcard() {
        return mSC;
    }

    public long getSlotNo() {
        return mSlotNo;
    }

    @Override
    public boolean equals(Object obj) {
        if(obj instanceof  SmartCardSeed) {
            SmartCardSeed scObj = (SmartCardSeed) obj;
            if (this.getSmartcard().getCardType().equals(scObj.getSmartcard().getCardType()) &&
                    this.getSlotNo() == scObj.getSlotNo())
                return true;
        }
        else{
            return false;
        }

        return false;
    }

}
